oclc/classify.py

```diff
@@ -32,6 +36,147 @@ class OCLC(object):
+    # defined in Contributor.
+    ROLE_MAPPING = {
+        "Author": Contributor.AUTHOR_ROLE,
+	"Translator": Contributor.TRANSLATOR_ROLE,
+	"Illustrator": Contributor.ILLUSTRATOR_ROLE,
+	"Editor": Contributor.EDITOR_ROLE,
+	"Unknown": Contributor.UNKNOWN_ROLE,
+	"Contributor": Contributor.CONTRIBUTOR_ROLE,
+	"Author of introduction": Contributor.INTRODUCTION_ROLE,
+	"Other": Contributor.UNKNOWN_ROLE,
+	"Creator": Contributor.AUTHOR_ROLE,
+	"Artist": Contributor.ARTIST_ROLE,
+	"Associated name": Contributor.ASSOCIATED_ROLE,
+	"Photographer": Contributor.PHOTOGRAPHER_ROLE,
+	"Compiler": Contributor.COMPILER_ROLE,
+	"Adapter": Contributor.ADAPTER_ROLE,
+	"Editor of compilation": Contributor.EDITOR_ROLE,
+	"Narrator": Contributor.NARRATOR_ROLE,
+	"Author of afterword, colophon, etc.": Contributor.AFTERWORD_ROLE,
+	"Performer": Contributor.PERFORMER_ROLE,
+	"Author of screenplay": Contributor.AUTHOR_ROLE,
+	"Writer of added text": Contributor.AUTHOR_ROLE,
+	"Composer": Contributor.COMPOSER_ROLE,
+	"Lyricist": Contributor.LYRICIST_ROLE,
+	"Author of dialog": Contributor.AUTHOR_ROLE,
+	"Film director": Contributor.DIRECTOR_ROLE,
+	"Actor": Contributor.ACTOR_ROLE,
+	"Musician": Contributor.MUSICIAN_ROLE,
+	"Filmmaker": Contributor.DIRECTOR_ROLE,
+	"Producer": Contributor.PRODUCER_ROLE,
+	"Director": Contributor.DIRECTOR_ROLE,
+    }
+
+    @classmethod
+    def parse_multiple(cls, author_string):
+        """Parse a list of people.
+
+        :return: A list of ContributorData objects.
+        """
+
+        # We start off assuming that someone with no explicit role
+        # is the primary author.
+        default_role = Contributor.PRIMARY_AUTHOR_ROLE
+        contributors = []
+        if not author_string:
+            return contributors
+        for author in author_string.split("|"):
+            contributor, default_role_used = cls.parse(author, default_role)
+            if contributor.roles:
+                if Contributor.PRIMARY_AUTHOR_ROLE in contributor.roles:
+                    # That was the primary author, or at least the
+                    # first author listed. If we see someone with no
+                    # explicit role after this point, assume they're
+                    # just a regular author.
+                    default_role = Contributor.AUTHOR_ROLE
+                elif not default_role_used:
+                    # We're dealing with someone whose role was
+                    # explicitly specified. If we see someone with no
+                    # explicit role after this point, it's probably
+                    # because their role is so minor as to not be
+                    # worth mentioning, not because it's so major that
+                    # we can assume they're an author.
+                    default_role = Contributor.UNKNOWN_ROLE
+            else:
+                # No explicit role was provided. Assign the default
+                # role.
+                contributor.roles = [default_role]
```

Comment:
If we assigned someone the primary author role at this point, should we change the default role to author here too?

Reply:
Yeah, I think we might need to make a little state machine here because the same logic applies when we're parsing multiple authors from subsequent <author> tags.